src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -115,32 +44,62 @@ module Make (Inputs : Intf.Main_inputs) = struct
             !"Unable to write mutant diff correctly as hashes are different:\n\
              \ %s. Hash of groundtruth %s Hash of actual %s"
             (Yojson.Safe.to_string
-               (Transition_frontier.Diff_mutant.key_to_yojson diff_mutant))
+               (Transition_frontier.Diff_mutant.key_to_yojson diff))
             (Transition_frontier.Diff_hash.to_string ground_truth_hash)
             (Transition_frontier.Diff_hash.to_string new_hash)
             ()
 
-  let listen_to_frontier_broadcast_pipe ~logger
-      (frontier_broadcast_pipe :
-        Transition_frontier.t option Broadcast_pipe.Reader.t) worker =
-    let%bind (_ : Transition_frontier.Diff_hash.t) =
-      Broadcast_pipe.Reader.fold frontier_broadcast_pipe
-        ~init:Transition_frontier.Diff_hash.empty
-        ~f:(fun acc_hash frontier_opt ->
-          match frontier_opt with
-          | Some frontier ->
-              Broadcast_pipe.Reader.fold
-                (Transition_frontier.persistence_diff_pipe frontier)
-                ~init:acc_hash ~f:(fun acc_hash diffs ->
-                  Deferred.List.fold diffs ~init:acc_hash
-                    ~f:(fun acc_hash (E mutant) ->
-                      write_diff_and_verify ~logger ~acc_hash worker frontier
-                        mutant ) )
-          | None ->
-              (* TODO: need to delete persistence once it get's back up *)
-              Deferred.return Transition_frontier.Diff_hash.empty )
+  let create ?directory_name ~logger max_buffer_capacity =
+    let worker = Worker.create ?directory_name ~logger () in
+    let reader, writer = Pipe.create () in
+    let buffer = Queue.create () in
+    let worker_thread =
+      Pipe.fold ~flushed:Pipe.Flushed.When_value_processed
+        ~init:Transition_frontier.Diff_hash.empty reader
+        ~f:(fun init_hash diff_pairs ->
+          Deferred.List.fold diff_pairs ~init:init_hash
+            ~f:(fun acc_hash
+               (Transition_frontier.Diff_mutant.E.Pair
+                 (diff, ground_truth_mutant))
+               ->
+              let%bind result =
+                write_diff_and_verify ~logger ~acc_hash worker
+                  (diff, ground_truth_mutant)
+              in
+              (* We would want the scheduler to run other jobs after computing a diff *)
+              Deferred.create @@ fun ivar -> Ivar.fill ivar result ) )
```

Comment:
You don't need to do this. You can just return the `result Deferred.t` from above. The bind will happen inside of the loop internal to `Deferred.List.fold`.